lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -964,6 +964,91 @@ bool firrtl::areTypesConstCastable(FIRRTLType destFType, FIRRTLType srcFType,
+  if (dstRefType == srcRefType)
+    return true;
+  if (dstRefType.getForceable() && !srcRefType.getForceable())
+    return false;
+
+  // Okay walk the types recursively.  They must be identical "structurally"
+  // with exception leaf (ground) types of destination can be uninferred
+  // versions of the corresponding source type. (can lose width information or
+  // become a more general reset type)
+
+  // NOLINTBEGIN(misc-no-recursion)
+  auto recurse = [&](auto &&f, FIRRTLBaseType dest,
+                     FIRRTLBaseType src) -> bool {
+    // Fast-path for identical types.
+    if (dest == src)
+      return true;
+
+    // Always passive inside probes, but for sanity assert this.
+    assert(dest.isPassive() && src.isPassive());
+
+    // Recurse through aggregates to get the leaves, checking
+    // structural equivalence re:element count + names.
+
+    if (auto destVectorType = dest.dyn_cast<FVectorType>()) {
+      auto srcVectorType = src.dyn_cast<FVectorType>();
+      return srcVectorType &&
+             destVectorType.getNumElements() ==
+                 srcVectorType.getNumElements() &&
+             f(f, destVectorType.getElementType(),
+               srcVectorType.getElementType());
+    }
+
+    if (auto destBundleType = dest.dyn_cast<BundleType>()) {
+      auto srcBundleType = src.dyn_cast<BundleType>();
+      if (!srcBundleType)
+        return false;
+      // (no need to check orientation, these are always passive)
+      auto destElements = destBundleType.getElements();
+      auto srcElements = srcBundleType.getElements();
+
+      return destElements.size() == srcElements.size() &&
+             llvm::all_of_zip(
+                 destElements, srcElements,
+                 [&](const auto &destElement, const auto &srcElement) {
+                   return destElement.name == srcElement.name &&
+                          f(f, destElement.type, srcElement.type);
+                 });
+    }
+
+    if (auto destEnumType = dest.dyn_cast<FEnumType>()) {
+      auto srcEnumType = src.dyn_cast<FEnumType>();
+      if (!srcEnumType)
+        return false;
+      auto destElements = destEnumType.getElements();
+      auto srcElements = srcEnumType.getElements();
+
+      return destElements.size() == srcElements.size() &&
+             llvm::all_of_zip(
+                 destElements, srcElements,
+                 [&](const auto &destElement, const auto &srcElement) {
+                   return destElement.name == srcElement.name &&
+                          f(f, destElement.type, srcElement.type);
+                 });
+    }
+
+    // Reset types can be driven by UInt<1>, AsyncReset, or Reset types.
+    if (dest.isa<ResetType>())
+      return src.isResetType();
+    // (but don't allow the other direction, can only become more general)
+
+    // src != dst, so this is only valid if destination is widthless version of
+    // source:
+    return dest.getBitWidthOrSentinel() == -1 && dest == src.getWidthlessType();
```

Comment:
I'm wondering if this should also allow for const decay. Currently this will error:
```
circuit Foo:
  module Foo:
    output p: Probe<UInt<1>>
    node n = UInt<1>(1)
    define p = probe(n)
```

Reply:
Oh what, since when are things automatically made const?

Almost left a comment noting that this intentionally doesn't try to support const decay, seems weird to have the const "equivalence" present here.  Literals being const and that not interacting well with consistently-typed refs is a new consideration.

Anything else not const in the source becomes const automatically?

Reply:
Yeah, the literal change was recently implemented but was always the plan (it's in the spec!). Currently it's just integer literals, if we ever get aggregate literals it'll apply to those too.

In terms of implementation, we could either do it here or have const-cast also support ref types I suppose.